modules/imgproc/src/subdivision2d.cpp

```diff
@@ -210,19 +210,137 @@ void Subdiv2D::swapEdges( int edge )
+    return val > eps ? 1 : val < -eps ? -1 : 0;
+}
+
+// the "meta point" predicate
+#undef MP
+#define MP(i) ((i) == 1 || (i) == 2 || (i) == 3)
+
+int Subdiv2D::counterClockwiseEx(int i, int j, int k) const
+{
+    if (MP(i) && MP(j) && MP(k)) {
+        return counterClockwise(vtx[i].pt, vtx[j].pt, vtx[k].pt);
+    }
+
+    if (MP(i) && MP(j) && !MP(k)) {
+        return counterClockwise(vtx[i].pt, vtx[j].pt, vtx[0].pt);
+    }
+
+    if (!MP(i) && !MP(j) && MP(k)) {
+        return !parallel(vtx[j].pt - vtx[i].pt, vtx[k].pt) ?
+                counterClockwise(vtx[0].pt, vtx[j].pt - vtx[i].pt, vtx[k].pt) :
+                counterClockwise(vtx[i].pt, vtx[j].pt, vtx[0].pt);
+    }
+
+    if (!MP(i) && !MP(j) && !MP(k)) {
+        return counterClockwise(vtx[i].pt, vtx[j].pt, vtx[k].pt);
+    }
+
+    return counterClockwiseEx(j, k, i);
+}
+
+int Subdiv2D::inCircleEx(int i, int j, int k, int l) const
+{
+    if (MP(i) && MP(j) && MP(k) && !MP(l)) {
+        return counterClockwiseEx(i, j, k);
+    }
+
+    if (MP(i) && !MP(j) && MP(k) && !MP(l)) {
+        if (!parallel(vtx[l].pt - vtx[j].pt, vtx[k].pt - vtx[i].pt)) {
+            return leftOf(vtx[l].pt - vtx[j].pt, vtx[0].pt, vtx[k].pt - vtx[i].pt);
+        } else {
+            double dl = distance(vtx[0].pt, vtx[l].pt);
+            double dj = distance(vtx[0].pt, vtx[j].pt);
+
+            if (abs(dl - dj) < FLT_EPSILON) {
+                return 0;
+            } else {
+                return (dl < dj && counterClockwiseEx(i, j, k) > 0) || (dl > dj && counterClockwiseEx(i, j, k) < 0) ? 1 : -1;
+            }
+        }
+    }
+
+    if (MP(i) && MP(j) && !MP(k) && !MP(l)) {
+        return -inCircleEx(i, k, j, l);
+    }
+
+    if (!MP(i) && !MP(j) && !MP(k) && MP(l)) {
+        if (!parallel(vtx[j].pt - vtx[i].pt, vtx[k].pt - vtx[i].pt)) {
+            return -counterClockwise(vtx[i].pt, vtx[j].pt, vtx[k].pt);
+        } else {
+            return !vtx[k].pt.inside(Rect2f(vtx[i].pt, vtx[j].pt)) ?
+                   leftOf(vtx[l].pt, vtx[0].pt, vtx[j].pt - vtx[i].pt) :
+                   leftOf(vtx[l].pt, vtx[0].pt, vtx[k].pt - vtx[j].pt);
+        }
+    }
+
+    if (!MP(i) && !MP(j) && !MP(k) && !MP(l)) {
+        return inCircle(vtx[i].pt, vtx[j].pt, vtx[k].pt, vtx[l].pt);
+    }
+
+    return -inCircleEx(j, k, l, i);
+}
+
+int Subdiv2D::rightOfEx(int k, int i, int j) const
+{
+    return counterClockwiseEx(k, j, i);
+}
+
+int Subdiv2D::rightOfEx(Point2f c, int i, int j) const
+{
+    if (!MP(i) && !MP(j)) {
+        return counterClockwise(c, vtx[j].pt, vtx[i].pt);
+    }
+
+    if (!MP(i) && MP(j)) {
+        return !parallel(c - vtx[i].pt, vtx[j].pt) ?
+               counterClockwise(vtx[0].pt, c - vtx[i].pt, vtx[j].pt) :
+               counterClockwise(vtx[i].pt, c, vtx[0].pt);
+    }
+
+    if (MP(i) && !MP(j)) {
+        return -rightOfEx(c, j, i);
+    }
+
+    return counterClockwise(vtx[j].pt, vtx[i].pt, vtx[0].pt);
+}
+
+// deprecated
```

Comment:
@vpisarev isRightOf(vtx[1..3].pt, ...) is meaningless despite rightOfEx(1..3, ...) is perfectly valid (strictly speaking the meta points have no coordinates). Is it possible to hide the vtx array under "private" modifier in future, so that accessors (getVertex(), edgeOrg, and edgeDst()) could ensure the points are not leaking out?